storage: reject range transfers when draining

asubiotto opened the pull request:
Draining stores now reject preemptive snapshots to avoid a possible loss
of quorum after the node shuts down.

`DrainLeases` and `IsDrainingLeases` have been renamed to `SetDraining`
and `IsDraining` to follow the general draining naming convention and because a
draining store does more than just drain leases now.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13601)
<!-- Reviewable:end -->

bdarnell commented:
Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/storage/store.go, line 963 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13601#-Kd27i818NXsLYxuvk4D:-Kd27i818NXsLYxuvk4E:b-u2g918) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2aac5115da919087d46f60dcfa1a7b1aa32b19c/pkg/storage/store.go#L963)):*
> ```Go
> // there may be active leases held by some of the Store's Replicas.
> // When called with 'false', returns to the normal mode of operation.
> func (s *Store) SetDraining(drain bool) error {
> ```

As we discussed at lunch, there are multiple kinds of draining (leases, replicas, clients, etc), Probably better to keep this `DrainLeases` unless we're going to adopt a new naming scheme in which "draining" unambiguously refers to leases. (it's also best to do renames like this in a separate commit from any functional change)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13601)*
<!-- Sent from Reviewable.io -->

Commit:
storage: rename DrainLeases to SetDraining

This change renames the method used to put a Store into draining mode as
well as all related methods/member variables. Previously all that was done
was to gracefully wait for all leases to expire, but future changes will
add more steps to this process that are unrelated to leases.

Commit:
storage: reject range transfers when draining

Draining stores now reject preemptive snapshots to avoid a possible loss
of quorum after the node shuts down.

asubiotto commented:
Review status: 0 of 7 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/storage/store.go, line 963 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13601#-Kd27i818NXsLYxuvk4D:-Kd6loXkiGVj_N5V1Swc:bba7j54) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2aac5115da919087d46f60dcfa1a7b1aa32b19c/pkg/storage/store.go#L963)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

As we discussed at lunch, there are multiple kinds of draining (leases, replicas, clients, etc), Probably better to keep this `DrainLeases` unless we're going to adopt a new naming scheme in which "draining" unambiguously refers to leases. (it's also best to do renames like this in a separate commit from any functional change)
</blockquote></details>

There are multiple kinds of draining, but in the context of a `Store` I'm not sure it makes sense to separate the draining of leases and replicas. This is because the more important distinction is draining for a restart or draining for a removal. You would want to stop accepting replicas in both cases to avoid a possible loss of quorum. I think I'll keep the rename (put it in its own commit) and later on add a boolean `forRemoval` that will transfer away replicas from a store.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13601)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 7 files at r1, 5 of 5 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13601#-:-Kd7OCHRf-55vU_t9yJ7:bnfp4nl)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13601#-:-KdDQHiT0RJot0SnPMVX:bnfp4nl)*
<!-- Sent from Reviewable.io -->
